### ISSUE TITLE
no-unexpected-multiline should check for `tagged template strings`

### ISSUE THREAD

**Issue by feross** [eslint/eslint#4169]
This code:

```js
let x = () => {}
`hello`
console.log(x)
```

will output:

```
undefined
```

I propose that the [no-unexpected-multiline](http://eslint.org/docs/rules/no-unexpected-multiline) rule also check for this case.

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by nzakas** [comment]
Can you explain more about what's happening here?

**Comment by feross** [comment]
```js
let x = () => {}
`hello`
console.log(x)
```

gets interpreted as:

```js
let x = () => {}`hello`
console.log(x)
```

In other words, the function is interpreted as the *tag* for purposes of the template string that follows. This isn't what the user wants if they're relying on ASI.

**Comment by feross** [comment]
More info about template strings: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/template_strings

**Comment by michaelficarra** [comment]
No, an arrow isn't allowed there. A template tag must be a call expression or member expression. Arrows are assignment expressions.

**Comment by feross** [comment]
Can you explain why:

```js
let x = () => {}
`hello`
console.log(x)
```

returns `undefined`?

**Comment by michaelficarra** [comment]
@feross Are you using babel? Babel doesn't parse arrows with the correct precedence. See https://github.com/babel/babel/issues/2154 and many other issues.

**Comment by feross** [comment]
No, I just tested directly in node 4.

**Comment by michaelficarra** [comment]
Then they are wrong too.

**Comment by feross** [comment]
Node just uses V8.

**Comment by michaelficarra** [comment]
Then V8 is wrong. That's just not how that program is parsed. Take a look at the grammar.

**Comment by mysticatea** [comment]
http://www.ecma-international.org/ecma-262/6.0/#sec-left-hand-side-expressions
There is not `[no LineTerminator here]` mark between MemberExpression and TemplateLiteral, different from http://www.ecma-international.org/ecma-262/6.0/#sec-return-statement

It looks not wrong to me, but not 100% sure.

**Comment by michaelficarra** [comment]
Arrows are not MemberExpressions. You'd need to wrap the arrow in parens for it to parse (in a conforming implementation).

**Comment by mysticatea** [comment]
On, you are right.
PrimaryExpression does not include Arrows. Though it includes FunctionExpression and ClassExpression....

**Comment by nzakas** [comment]
@feross per discussion, I'd advise you to open an issue on V8.

**Comment by feross** [comment]
Opened a bug on V8: https://code.google.com/p/v8/issues/detail?id=4497

**Comment by mysticatea** [comment]
@michaelficarra BTW, if there is PrimaryExpression instead of ArrowFunctionExpression, should `no-unexpected-multiline` catch that? e.g.

```js
let x = function() {}
`hello`
```

**Comment by feross** [comment]
@mysticatea Good point. I think it should. That's what the ASI users will expect, and that's what that rule was designed to catch.

**Comment by michaelficarra** [comment]
@mysticatea Yeah, that seems appropriate. By the way, a coworker of mine tried to get a `[no LineTerminator here]` put in back in February, but @zenparsing thought it would be inconsistent with calls and member accesses, and it didn't receive any other committee feedback :disappointed:.

**Comment by feross** [comment]
@michaelficarra @nzakas can we re-open this issue so we can track improvement of the rule as discussed in @mysticatea's last comment?

**Comment by michaelficarra** [comment]
@feross I'd say a separate issue is more appropriate.

**Comment by feross** [comment]
@michaelficarra Sure thing: https://github.com/eslint/eslint/issues/4210
